Enable more MongoDB Micrometer metrics

vkn opened the pull request:
Currently only connection pool micrometer metrics are supported. This PR enables all available metrics based on `io.micrometer.core.instrument.binder.mongodb.MongoMetricsCommandListener`

All unit tests in mongodb-client extension are based on microprofile metrics, so it was not possible to add tests for micrometer, without modifying dependencies and existing tests. However I've added an assertion in MongodbPanacheResourceTest.java in integration-tests/mongodb-panache  module

gsmet commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:
```diff
@@ -124,6 +124,17 @@ AdditionalIndexedClassesBuildItem includeMongoCommandListener(MongoClientBuildTi
+    AdditionalIndexedClassesBuildItem includeMongoCommandMetricListener(
+            MongoClientBuildTimeConfig buildTimeConfig,
+            Optional<MetricsCapabilityBuildItem> metricsCapability) {
+        return metricsCapability
+                .filter(cap -> buildTimeConfig.metricsEnabled && cap.metricsSupported(MetricsFactory.MICROMETER))
+                .map(supported -> new AdditionalIndexedClassesBuildItem(
+                        MongoClientRecorder.getMicrometerCommandListenerClassName()))
+                .orElseGet(AdditionalIndexedClassesBuildItem::new);
```
Any chance you could implement this with simple ifs (with early returns for edge cases)?

Because this is a bit hard to follow and I'm worried we will regret it when fixing a bug one early morning without coffee :).

vkn replied:
@gsmet pushforced the change

gsmet reviewed (commented):
Thanks for the PR! I added a small comment inline.

I will defer to Loïc and Bruno to check the feature.

vkn committed:
Enable mongodb micrometer metrics

brunobat reviewed (changes requested):
It looks ok to me but could you please add a test entry here: https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/MongoMetricsTest.java#L24

vkn commented:
> It looks ok to me but could you please add a test entry here:
> 
> https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/MongoMetricsTest.java#L24

@brunobat the test is based on microprofile metrics. to add test for the micrometer metrics, we would have to refactor tests and create separate tests for micrometer and microprofile metrics. or do you see an easier way

brunobat commented:
> > It looks ok to me but could you please add a test entry here:
> > https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/MongoMetricsTest.java#L24
> 
> @brunobat the test is based on microprofile metrics. to add test for the micrometer metrics, we would have to refactor tests and create separate tests for micrometer and microprofile metrics. or do you see an easier way

yeah, sorry, didn't look at the imports... Then It's worse than I thought.
Can you please add a test class for Micrometer under the deployment module?

vkn commented:
> > > It looks ok to me but could you please add a test entry here:
> > > https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/MongoMetricsTest.java#L24
> > 
> > 
> > @brunobat the test is based on microprofile metrics. to add test for the micrometer metrics, we would have to refactor tests and create separate tests for micrometer and microprofile metrics. or do you see an easier way
> 
> yeah, sorry, didn't look at the imports... Then It's worse than I thought. Can you please add a test class for Micrometer under the deployment module?

@brunobat The deployment module pom has `quarkus-smallrye-metrics-deployment` dependency. so microprofile metrics are on the class path and there is no way to add micrometer metrics, unless you agree to remove it from the pom xml and add the dependency in QuarkusUnitTest programmatically. The existing test have to be then adjusted and not import microprofile metrics, as far as I can tell

loicmathieu commented:
>  The deployment module pom has quarkus-smallrye-metrics-deployment dependency. so microprofile metrics are on the class path and there is no way to add micrometer metrics,

Mongodb metrics support preclude our switch to Micrometer by default that's why the test is based on Microprofile metrics, I'm OK to switch to test to Micrometer.

@vkn thanks a lot for your contribution. MongoDB could emit a lot of commands, I wonder if this listener should be added always of if we could be able to disable it for performance reason.

loicmathieu commented on extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java:
```diff
@@ -99,6 +100,10 @@ public ConnectionPoolListener get() {
         };
     }
 
+    public static String getMicrometerCommandListenerClassName() {
+        return MicrometerCommandListener.class.getName();
+    }
+
```
As this is used in a single place you can directly add the class name at the use site.
This would also avoid loading the class at runtime if not used wich would be better.

vkn replied:
> As this is used in a single place you can directly add the class name at the use site. This would also avoid loading the class at runtime if not used wich would be better.

micrometer wasn't on class path in deployment module, that's why I've put it there. I'll add optional dependency and move the code to the deployment

loicmathieu reviewed (dismissed):
As other already said, a test would be great

vkn commented:
@brunobat @loicmathieu 
I've changed the tests to be based on micrometer. `MongoMetricsTest ` verifies the new `mongodb.driver.commands` metrics are available. Please have a look

brunobat commented on extensions/mongodb-client/deployment/pom.xml:
```diff
@@ -63,8 +63,8 @@
             <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-smallrye-metrics-deployment</artifactId>
-            <scope>test</scope>
+            <artifactId>quarkus-micrometer-registry-prometheus-deployment</artifactId>
```
Can you make it with just the `quarkus-micrometer` extension?

vkn replied:
@brunobat I get:
```
[ERROR] Rule 4: io.quarkus.enforcer.BansRuntimeDependency failed with message:
[ERROR] 1 illegal runtime dependencies found that have to be replaced with their -deployment counterparts:
[ERROR]         quarkus-micrometer
```

brunobat replied:
you need the `quarkus-micrometer-deployment`

brunobat replied:
I mean, replace `quarkus-micrometer-registry-prometheus-deployment` with `quarkus-micrometer-deployment`.
I'm not sure if it will cause you an error on the tests, if so `quarkus-micrometer-registry-prometheus-deployment` can be there but as a test dependency.
The reason is that the `quarkus-micrometer-registry-prometheus-deployment` is a particular telemetry output lib for prometheus, there are other implementations like otlp and we should avoid depending directly on them o the main code.

vkn replied:
Makes sense. I did exactly that. Still need to fix imports sorting. Then it should be ok

brunobat replied:
Thanks!

vkn committed:
Migrate tests to micrometer metrics

brunobat reviewed (approved):
Looks good to me.
Thanks @vkn !

gsmet reviewed (changes requested):
Thanks for the adjustment. I requested a small change as we shouldn't produce an empty `AdditionalIndexedClassesBuildItem` when we don't want to produce anything.

Once this is fixed, it's all good to go.

I can do it if you prefer but I thought you might be interested in doing it yourself. Let me know!

vkn committed:
Avoid producing empty AdditionalIndexedClassesBuildItem

Even if it doesn't complain right now,
that's certainly something that could break later

vkn commented on extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:
```diff
@@ -124,6 +125,21 @@ AdditionalIndexedClassesBuildItem includeMongoCommandListener(MongoClientBuildTi
+            MongoClientBuildTimeConfig buildTimeConfig,
+            Optional<MetricsCapabilityBuildItem> metricsCapability) {
+        if (!buildTimeConfig.metricsEnabled) {
+            return new AdditionalIndexedClassesBuildItem();
```
@gsmet  I've adjusted the code as you requested. Thanks!
